### PR TITLE
Fix processing of account.application.deauthorized events

### PR DESCRIPTION
**Pull request by ticosax**
don't try to fetch the event as we lost access to remote account

### CODE REVIEW

File: pinax/stripe/webhooks.py [hunk 1]
```diff
@@ -144,8 +144,30 @@ class AccountApplicationDeauthorizeWebhook(Webhook):
...
+        stripe_account_id = self.event.webhook_message["account"]
+        self.stripe_account = models.Account.objects.filter(stripe_id=stripe_account_id).first()
+        try:
+            stripe.Event.retrieve(
+                self.event.stripe_id,
+                stripe_account=stripe_account_id,
+            )
+        except stripe.error.PermissionError as exc:
+            if stripe_account_id not in str(exc):
+                raise exc
```
Comment (blueyed) [inline pinax/stripe/webhooks.py]: This should be more specific, and needs to cover the re-raise.

File: pinax/stripe/webhooks.py [hunk 2]
```diff
@@ -144,8 +144,30 @@ class AccountApplicationDeauthorizeWebhook(Webhook):
     name = "account.application.deauthorized"
     description = "Occurs whenever a user deauthorizes an application. Sent to the related application only."
 
+    def validate(self):
+        """
+        Validate incoming events.
+
+        We don't fetch remote the event as we lost access to the account where the event belongs.
```
Comment (blueyed) [inline pinax/stripe/webhooks.py]: doc needs to be adjusted.

File: pinax/stripe/webhooks.py [hunk 3]
```diff
@@ -144,8 +144,30 @@ class AccountApplicationDeauthorizeWebhook(Webhook):
     name = "account.application.deauthorized"
     description = "Occurs whenever a user deauthorizes an application. Sent to the related application only."
 
+    def validate(self):
+        """
+        Validate incoming events.
```
Comment (blueyed) [inline pinax/stripe/webhooks.py]: *"Specialized validation of incoming events."

File: pinax/stripe/webhooks.py [hunk 4]
```diff
@@ -144,8 +145,30 @@ class AccountApplicationDeauthorizeWebhook(Webhook):
...
+                stripe_account=stripe_account_id,
+            )
+        except stripe.error.PermissionError as exc:
+            if not(stripe_account_id in str(exc) and obfuscate_secret_key(settings.PINAX_STRIPE_SECRET_KEY) in str(exc)):
+                raise exc
+            self.event.valid = True
+            self.event.validated_message = self.event.webhook_message
+            self.event.stripe_account = self.stripe_account
+        else:
+            raise ValueError("The remote account is still valid. This might be a hostile event.")
```
Comment (blueyed) [inline pinax/stripe/webhooks.py]: changed the message, please pull.

File: pinax/stripe/tests/test_webhooks.py [hunk 5]
```diff
@@ -602,14 +602,42 @@ def test_process_webhook(self, SyncMock, RetrieveMock):
...
+                "account": self.account.stripe_id}
         event = Event.objects.create(
             kind=AccountApplicationDeauthorizeWebhook.name,
-            webhook_message={},
-            valid=True,
-            processed=False
+            webhook_message=data,
+        )
+        RetrieveMock.side_effect = stripe.error.PermissionError(
+            "The provided key 'sk_test_********************abcd' does not have access to account 'acc_aa'")
```
Comment (blueyed) [inline pinax/stripe/tests/test_webhooks.py:614]: That wasn't the full message, was it?

File: pinax/stripe/webhooks.py [hunk 6]
```diff
@@ -144,8 +145,34 @@ class AccountApplicationDeauthorizeWebhook(Webhook):
...
+        Specialized validation of incoming events.
+
+        We try to retrieve the event:
+         - in case of PermissionError exception, everything is perfectly normal.
+           It means the account has been deauthorized.
+         - In case no exception has been caught, then, most likely, the event has been forged
+           to make you believe the account has been disabled despite it is still functioning.
+        """
+        stripe_account_id = self.event.webhook_message["account"]
+        self.stripe_account = models.Account.objects.filter(stripe_id=stripe_account_id).first()
```
Comment (blueyed) [inline pinax/stripe/webhooks.py:159]: `.get()`?

File: pinax/stripe/webhooks.py [hunk 7]
```diff
@@ -144,8 +145,34 @@ class AccountApplicationDeauthorizeWebhook(Webhook):
...
+                stripe_account=stripe_account_id,
+            )
+        except stripe.error.PermissionError as exc:
+            if not(stripe_account_id in str(exc) and obfuscate_secret_key(settings.PINAX_STRIPE_SECRET_KEY) in str(exc)):
+                raise exc
+            self.event.valid = True
+            self.event.validated_message = self.event.webhook_message
+            self.event.stripe_account = self.stripe_account
+        else:
+            raise ValueError("The remote account still valid. this might be an hostile event")
```
Comment (blueyed) [inline pinax/stripe/webhooks.py:172]: Tried twice to fix this: https://github.com/lock8/pinax-stripe/commit/c962eaf

File: pinax/stripe/tests/test_webhooks.py (hunk 5 above)
Comment (ticosax) [reply pinax/stripe/tests/test_webhooks.py:614]: yes

File: pinax/stripe/webhooks.py (hunk 6 above)
Comment (ticosax) [reply pinax/stripe/webhooks.py:159]: Do we want to blow up in case we are not in sync ?

File: pinax/stripe/webhooks.py (hunk 7 above)
Comment (ticosax) [reply pinax/stripe/webhooks.py:172]: Pushing to my fork is fine when I don't work on it.
it means I have time to read the notification.
thank you anyway

File: pinax/stripe/webhooks.py (hunk 6 above)
Comment (ticosax) [reply pinax/stripe/webhooks.py:159]: in lock8/pinax-stripe#35

File: pinax/stripe/webhooks.py (hunk 7 above)
Comment (blueyed) [reply pinax/stripe/webhooks.py:172]: Sure.  There is `--force-with-lease` for git-push to only force-push is what you expect it to be.

**Pull request merged**